Add sections on comments and error messages in coding guide (#1660)
PlasmaPy/PlasmaPy: docs/contributing/code_guide.rst

> PlasmaPy fork on GitHub.  After a pull request is merged into the

Reviewer: The coding guide is getting rather long.  I'm thinking it might be time to add a ``.. contents:` to the top of the page.  Thoughts?

Author: Yes!

Author: Addressing it in #1662 for not only the coding guide, but also the other pages in the contributor guide.

code, you may delete the branch you created for that pull request.

> Comments

Reviewer: I wonder if establishing a "golden rule" would be useful here.  Maybe even put it into an admonition.  I personally get lost in a list of guide points, so having a golden rule to check against is what I find most useful.  The golden rule I tend to tell people goes something like:

> Your code should read like a book, that is the full meaning of your functionality should be attainable by reading the code.  A comment is only used when code itself can not communicate the full detailed meaning.

I'm sure there's a better way of phrasing this, but I think it covers all the points you say below.  I think pairing the "golden rule" with your points as examples would be overall more effective.

Thoughts?

Author: Good idea!  I'll add something like this in, likely after I get back from visiting family.

========

A well-placed and well-written comment can prevent future frustrations.
However, comments are not inherently good. As code evolves, an
unmaintained comment may become outdated, or get separated from the
section of code that it was meant to describe. Cryptic and obsolete
comments may end up confusing contributors. In the worst case, an
unmaintained comment may contain inaccurate or misleading information
(hence the saying that "a comment is a lie waiting to happen").

* Refactor code to make it more readable, rather than explaining how it
  works :cite:p:`wilson:2014`.

* Instead of using a comment to define a variable, rename the variable
  to encode its meaning and intent.  For example, code like:

  .. code-block:: python

     # collision frequency
     nu = 1e6 * u.s ** -1

  could be changed to:

  .. code-block:: python

>      collision_frequency = 1e6 * u.s ** -1  # no comment needed!

Author: Oh good point!

* Use comments to communicate information that you wish you knew before
  starting to work on a particular section of code, including
  information that took some time to learn.

* Use comments to communicate information that the code cannot,
  such as why an alternative approach was *not* taken.

* When possible, include enough contextual information that the comment
  will make sense even if it is displaced from the code it was

>   originally describing.

Reviewer: I'm not sure how to visualize this point.  Is there a good example.  I think it's a little confusing without more context. (Sorry, I couldn't resist saying that last bit.)

Author: I decided to delete this point since there is already another one that says to provide context.

* Use comments to include references to books or articles that describe
  the equation, algorithm, or software design pattern that is being
  implemented. Even better, include these references in docstrings.

* Provide enough contextual information in the comment for a new user
  to be able to understand it.

* Remove commented out code before merging a pull request.

> * When updating code, be sure to update the comments too!

Reviewer: ```suggestion
* When updating code, be sure to review and update, if necessary, associated comments too!
```

> * When a comment is used as the header for a section of code, consider

Reviewer: You're going to make me rage quit with this point. 😂 I'd like to see some kind of comment in here that there's a "subjective" balancing act with the `extract function refactoring pattern`.  Overall it's good, but when used on a too fine of a scale it can significantly reduce code readability.

Author: That's a good point!  This is a good place to put in a tradeoff analysis and how there are different factors we need to balance.

Author: Added in a discussion on this below.

         ...
         return calibrated_data

     def normalize_data(data):
         ...
         return normalized_data

     def analyze_experiment(data):
         calibrated_data = calibrate_data(data)
         normalized_data = normalize_data(calibrated_data)

  This refactoring strategy is appropriate for long functions where the
  different steps can be cleanly separated from each other. This pattern
  leads to functions that are shorter, more focused, more reusable, and
  easier to test. The original function no longer includes
  implementation details, and thus gives a high level view of what the
  function is doing.

  .. hint::

     The `extract function refactoring pattern`_ might not be
     appropriate if the different sections of code are intertwined with
     each other (e.g., if both sections require the same intermediate
     variables). An alternative in such cases would be to create a class
     instead.

Error messages
==============

Error messages are a vital but underappreciated form of documentation.
A good error message can help someone pinpoint the source of a problem
in seconds, while a cryptic or missing error message can lead to hours

> of frustration.

Reviewer: I had a `sphinx` error like this yesterday.  It didn't even tell me what file the error was in. Ahhhh... 😤

Author: YEP.

> * Use error messages to indicate the source of the problem while
>   providing enough information for the user to fix it. When possible,
>   make it clear what the user should do next.

Reviewer: ```suggestion
* Use error messages to indicate the source of the problem while
  providing enough information for the user to troubleshoot it. When
  possible, make it clear what the user should do next.
```

Only because errors can be the side effect of another issue.

> * Include diagnostic information when appropriate.  If an occurred at
>   a single point in an array operation, including the coordinates of
>   where the error happened can help the user better understand the cause
>   of the error.

Reviewer: ```suggestion
* Include diagnostic information when appropriate.  For example, if an
  error occurred at a single index in an array operation, then including
  the index where the error happened can help the user better understand
  the cause of the error.
```

Author: Thanks!  I was having trouble wording this.

* Write error messages that are concise when possible, as users often
  skim or skip long error messages.

* Avoid including information that is irrelevant to the source of the
  problem.

* Write error messages in language that is plain enough to be
  understandable to someone who is undertaking their first research
  project.

  - If necessary, technical information may be placed after a plain
    language summary statement.

  - Alternatively, an error message may reference a docstring or a page
    in the narrative documentation.

* Write error messages that are friendly, supportive, and helpful. Error
  message should never be condescending or blame the user.

* The exceptions raised by a method should be described in the

>   method's docstring.  Documenting exceptions makes it easier for

Reviewer: Do we have a doc guide rule on this?  This is one of the most difficult parts of writing a docstring.  And, I often find this part of the docstring is like a comment, always out of date.

Author: Hm...perhaps we could emulate the [numpydoc section](https://numpydoc.readthedocs.io/en/latest/format.html#raises) on this, and say that the `Raises` section should be used judiciously and only include non-obvious exceptions.  

- [ ] Do ↑
- [ ] Update the doc guide section on this too

> And, I often find this part of the docstring is like a comment, always out of date.

All too true...

Author: Updated this, and moved it to the doc guide as a means of procrastinating packing.
